brainiak/utils/fmrisim.py

```diff
@@ -1595,6 +1603,66 @@ def _generate_noise_temporal_drift(trs,
         phase = (timepoints / (trs - 1) * cycles * 2 * np.pi) + phaseshift
         noise_drift = np.sin(phase)
 
+    elif basis == 'cos_power_drop':
+
+        # Make a vector counting each TR
+        timepoints = np.linspace(0, trs - 1, trs) * tr_duration
+
+        # Specify the other timing information
+        duration = trs * tr_duration
+
+        # How bases do you have?
+        basis_funcs = int(np.floor(2 * duration))
+
+        noise_drift = np.zeros((timepoints.shape[0], basis_funcs))
+        for basis_counter in list(range(1, basis_funcs + 1)):
+            # What steps do you want to take for this basis function
+            random_phase = np.random.rand() * np.pi * 2
+
+            timepoint_phase = (timepoints / duration * np.pi * basis_counter)
+
+            # In radians, what is the value for each time point
+            timepoints_basis = timepoint_phase + random_phase
+
+            # Store the drift from this basis func
+            noise_drift[:, basis_counter - 1] = np.cos(timepoints_basis)
+
+        def power_drop(r, L, F):
+            # Function to return the drop rate for the power of basis functions
+            # In other words, how much should the weight of each basis function
+            # reduce in order to make the power you retain be above a 0.99
+            # r is the power reduction rate which should be between 0 and 1
+            # L is the duration of the run in seconds
+            # F is period of the cycle in seconds
+
+            percent_retained = 0.99  # What is the percentage of drift retained
+            numerator = 1 - r ** (4 * L / F)  # Power of this period
+            denominator = 1 - r ** (4 * L)  # Power of all periods
+
+            # Calculate the retained power
+            power_drop = abs((numerator / denominator) - percent_retained)
+            return power_drop
+
+        # Solve for power reduction rate.
+        # This assumes that r is between 0 and 1
+        # Takes the duration and period as arguments
+        sol = optimize.minimize_scalar(power_drop,
+                                       bounds=(0, 1),
+                                       method='Bounded',
+                                       args=(duration, period))
+
+        # Pull out the solution
+        r = sol.x
+
+        # Weight the basis functions based on the power drop off
+        basis_weights = r ** np.arange(basis_funcs)
```

Comment:
According to this line, I feel the `basis_funcs` is something that ranges from 1 to the *index* of the Nyquiest frequency. Assuming it is true, then now I suppose what you are doing is to let the weight of each frequency component to decay exponentially. At each consecutive frequency, you decay by `r`.